lib/flow/Advection.cpp

```diff
@@ -42,14 +42,108 @@ Advection::CheckReady() const
     return 0;
 }
 
+int
+Advection::AdvectSteps( Field* velocity, float deltaT, size_t maxSteps, ADVECTION_METHOD method )
+{
+    int ready = CheckReady();
+    if( ready != 0 )
+        return ready;
+    // The particle advection process can be parallelized per particle
+    // Each stream represents a trajectory for a single particle
+    #pragma omp parallel
+    #pragma omp for
+    for(size_t streamIdx = 0; streamIdx < _streams.size(); streamIdx++)  // Process one stream at a time
+    {
+        auto& s = _streams.at(streamIdx);
+        size_t numberOfSteps = 0;
+        while(numberOfSteps < maxSteps)
+        {
+            // Check if the particle is inside of the volume.
+            // Also wrap it along periodic dimensions if enabled.
+            if( !velocity->InsideVolumeVelocity( s.back().time, s.back().location ) )
+            {
+                auto& p0 = s.back();
+                // Attempt to apply periodicity
+                bool locChanged = false;
+                auto loc = p0.location;
+                for( int i = 0; i < 3; i++ )    // correct coordinates in each periodic dimension
+                {
+                    if( _isPeriodic[i] )
+                    {
+                        loc[i] = _applyPeriodic( loc[i], _periodicBounds[i][0], _periodicBounds[i][1] );
+                        locChanged = true;
+                    }
+                }
+
+                if( !locChanged )   // no dimension is periodic
+                    break;       // skip this particle, since it's out of the volume
```

Comment:
Which loop does this break statement break?

Reply:
The statement breaks the while loop which iterates over the maximum steps for a single particle.
If the particle goes out of domain, we no longer need to continue advecting the while loop.
In a serial setting this will enable proceeding to the next particle.

Reply:
I see, yeah, I understand now. Thanks for the explanation!